## 4. Machine Learning Development

Importamos los .csv extraídos de los procesos anteriores para poner a trabajar al modelo.

In [2]:
import pandas as pd

In [3]:

y_train = pd.read_csv('/workspaces/EDA_3/data/processed/medical_insurance_y_train.csv')
y_test = pd.read_csv('/workspaces/EDA_3/data/processed/medical_insurance_y_test.csv')

X_train_norm = pd.read_csv('/workspaces/EDA_3/data/interim/medical_insurance_X_train_std.csv')
X_test_norm = pd.read_csv('/workspaces/EDA_3/data/interim/medical_insurance_X_test_std.csv')

In [4]:
from sklearn.feature_selection import f_regression, SelectKBest

In [5]:
pred_nums = []
for i in range(len(X_train_norm.columns)):
    pred_nums.append(i)

In [6]:
# X_train_sel = []
# X_test_sel = []
# for number in pred_nums:
#     selection_model = SelectKBest(f_regression, k = number)
#     selection_model.fit(X_train_norm, y_train)
#     ix = selection_model.get_support()
#     X_train_sel.append(pd.DataFrame(selection_model.transform(X_train_norm), columns = X_train_norm.columns.values[ix]))
#     X_test_sel.append(pd.DataFrame(selection_model.transform(X_test_norm), columns = X_test_norm.columns.values[ix]))
    # X_train_sel["charges"] = y_train
    # X_test_sel["charges"] = y_test    
    # train_data = X_train_sel
    # test_data = X_test_sel
    # X_train = train_data.drop(["charges"], axis = 1)
    # y_train = train_data["charges"] 
    # X_test = test_data.drop(["charges"], axis = 1)
    # y_test = test_data["charges"]
    
    # from sklearn.linear_model import LinearRegression

    # model = LinearRegression()
    # model.fit(X_train, y_train)
    # y_pred = model.predict(X_test)
    # from sklearn.metrics import mean_squared_error, r2_score

    # print(f"Error cuadrático medio: {mean_squared_error(y_test, y_pred)}")
    # print(f"Coeficiente de determinación: {r2_score(y_test, y_pred)}")



In [7]:
selection_model = SelectKBest(f_regression, k = 5)
selection_model.fit(X_train_norm, y_train)
ix = selection_model.get_support()
X_train_sel = pd.DataFrame(selection_model.transform(X_train_norm), columns = X_train_norm.columns.values[ix])
X_test_sel = pd.DataFrame(selection_model.transform(X_test_norm), columns = X_test_norm.columns.values[ix])

/home/vscode/.local/lib/python3.11/site-packages/sklearn/utils/validation.py:1183: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [8]:
X_train_sel

0         1         2         3         4
0     0.976695 -1.000936  0.881367  0.749696  0.494439
1    -1.516009 -1.000936 -0.543716  1.582777 -2.022493
2     1.546456  0.999065  0.874754 -0.916468 -2.022493
3    -0.376487 -1.000936 -1.912589  1.582777  0.494439
4     0.691814 -1.000936 -0.106437 -0.916468  0.494439
...        ...       ...       ...       ...       ...
1064 -0.732588 -1.000936  1.365763  1.582777  0.494439
1065  1.688896  0.999065  1.024372 -0.916468  0.494439
1066 -1.017468  0.999065 -0.798313 -0.916468  0.494439
1067  0.691814  0.999065 -1.331479 -0.916468  0.494439
1068  1.760116  0.999065  1.420320 -0.083386  0.494439

[1069 rows x 5 columns]

In [9]:
# for train_data, y_train in zip(X_train_sel[1:], y_train):
#     if not isinstance(train_data, pd.DataFrame) or train_data.empty:
#         continue  # Skip this iteration if the element is not a non-empty DataFrame
    
#     X_train = train_data.drop(["charges"], axis=1)
#     y_train = train_data["charges"]

#     from sklearn.linear_model import LinearRegression
#     model = LinearRegression()
#     model.fit(X_train, y_train)
    
#     # Assuming X_test_sel is similar to X_train_sel
#     for test_data, y_test in zip(X_test_sel[1:], y_test):
#         if not isinstance(test_data, pd.DataFrame) or test_data.empty:
#             continue  # Skip this iteration if the element is not a non-empty DataFrame

#         X_test = test_data.drop(["charges"], axis=1)
#         y_test = test_data["charges"]

#         y_pred = model.predict(X_test)

#         from sklearn.metrics import mean_squared_error, r2_score
#         print(f"Error cuadrático medio: {mean_squared_error(y_test, y_pred)}")
#         print(f"Coeficiente de determinación: {r2_score(y_test, y_pred)}")


In [10]:
X_train_sel["charges"] = y_train
X_test_sel["charges"] = y_test 

Exportamos la selección de las 5 mejores variables con las que trabajaremos el modelo

In [11]:
X_train_sel.to_csv("/workspaces/EDA_3/data/interim/medical_insurance_X_train_sel.csv", index=False)
X_test_sel.to_csv("/workspaces/EDA_3/data/interim/medical_insurance_X_test_sel.csv", index=False)

In [12]:
train_data = X_train_sel
test_data = X_test_sel
train_data.head()

0         1         2         3         4      charges
0  0.976695 -1.000936  0.881367  0.749696  0.494439  11163.56800
1 -1.516009 -1.000936 -0.543716  1.582777 -2.022493  18223.45120
2  1.546456  0.999065  0.874754 -0.916468 -2.022493  46599.10840
3 -0.376487 -1.000936 -1.912589  1.582777  0.494439   6753.03800
4  0.691814 -1.000936 -0.106437 -0.916468  0.494439   8988.15875

In [13]:
X_train = train_data.drop(["charges"], axis = 1)
y_train = train_data["charges"]
X_train.head()

0         1         2         3         4
0  0.976695 -1.000936  0.881367  0.749696  0.494439
1 -1.516009 -1.000936 -0.543716  1.582777 -2.022493
2  1.546456  0.999065  0.874754 -0.916468 -2.022493
3 -0.376487 -1.000936 -1.912589  1.582777  0.494439
4  0.691814 -1.000936 -0.106437 -0.916468  0.494439

In [14]:
X_test = test_data.drop(["charges"], axis = 1)
y_test = test_data["charges"]
X_test.head()

0         1         2         3         4
0 -0.518927  0.999065  2.638749  0.749696  0.494439
1 -1.159908 -1.000936 -0.424683 -0.916468  0.494439
2  0.122054  0.999065  0.601972 -0.083386  0.494439
3  0.763034  0.999065  0.600318  0.749696 -2.022493
4  1.546456  0.999065  0.914432  1.582777  0.494439

In [15]:
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [16]:
print(f"Intercept (a): {model.intercept_}")
print(f"Coefficients (b): {model.coef_}")

Intercept (a): 13056.014647071092
Coefficients (b): [ 3541.14386216    59.09064291  1953.36365645   388.10075381
 -9402.14338583]


In [17]:
y_pred = model.predict(X_test)
y_pred

array([12074.05793548,  3055.43368303, 10041.97637018, 36296.42744771,
       16342.97649272, 29239.3209985 ,  2111.52815234, 14840.14411257,
        6534.73148334, 30448.17132866, 10030.510156  ,  6261.56333801,
       11395.71188275,  6759.87638705, 18042.23972806, 11127.58962741,
       15277.52736546,  5822.07151867,  3459.16398079, 30516.5969392 ,
         148.07925351,  4561.06841619, 13559.94178379,  3290.25583334,
       16109.31089706, 14563.81360667,  1307.10752905, 38772.5413421 ,
        5589.11413632, 10619.15532514,  3751.98165014,  6954.91935946,
        6951.40879315, 12456.42169182, 11082.88303507, 28861.52104589,
       26412.96140073,  8831.05421612, 13014.19296484,  8721.24263835,
       39358.16843331,  7681.18368985, 11780.72968483,  5901.63121114,
       26948.65064973, 12998.86235603,  7234.0382131 , 32561.15852559,
        4385.80725038, 34952.30878635, 34206.54690536, 11672.41755454,
       10878.06377051,  6799.11075419,  7582.49794652, 37794.28906972,
      

In [19]:
from sklearn.metrics import mean_squared_error, r2_score
import math
print(f"Error cuadrático medio: {math.sqrt(mean_squared_error(y_test, y_pred))}")
print(f"Coeficiente de determinación: {r2_score(y_test, y_pred)}")

Error cuadrático medio: 6282.739994993724
Coeficiente de determinación: 0.7673264668767327


## *Conclusión:*

Obtenemos un R2 del 0.77 lo cual no resulta del todo óptimo. 

Revisando outliers y variables tampoco encontramos nada que podamos variar en pro de nuestro modelo por lo tanto nos quedamos con este resultado. 

Concluimos que no podemos mejorar mucho más un modelo con estos datos proporcionados, loc cuál puede ser debido a pocas correlaciones fuertes con el target y a su vez con el escaso número de columnas.